### ISSUE TITLE
Increasing memory during study leading to OOM

### ISSUE THREAD

**Issue by jmakov** [optuna/optuna#3050]
Hi. First, thanks for this project! A couple of observations and I'm not sure if it's a memory leak or it's just how Optuna works:

1. study runs, OOM: I'm distributing optuna workers with ray (code below) and am observing increasing memory consumption of Optuna workers while the study runs (increases ~400MB for 20k trials per Optuna process). This becomes problematic and nodes are crashing with OOM. The objective function run takes about 500MB RAM (and this is constant) and for some reason the memory amount (of the optimization process) increases (also with calling gc.collect, so we assume it's connected with Optuna). Is this as intended or is there a memory leak?
2. study resumes, OOM: Studies are created with `load_if_exists=True`. If we run a study from beginning (all previous DBs erased)  to e.g. 100k trials, the memory usage is e.g. 3/4 of node's available RAM. But if I interrupt the optimization processes on the node (ctrl+c, processes end) and start again (as said, with `load_if_exists=True`), then the memory increases til OOM. So I can only resume from e.g. 50k trials but not from e.g. 100k (although, as said, the study ran w/o problems till e.g. 100k trials). 
3. study resumes, OOM earlier with specific samplers: For samplers like `CmaEsSampler(use_separable_cma=True)` we get OOM already with e.g. 50k trials which leaves us to believe that some samplers need more memory to resume. Is this behavior as intended?
4. high network traffic: With the default sampler we observe 50-90Mbps of network traffic between PostgreSQL and the node. With `CmaEsSampler(use_separable_cma=True)` about 120Mbps (constant traffic). Our objective/tunable function takes about 1s to compute. Is such high network traffic expected?

### Environment
- Optuna version: 2.10.0
- Python version: 3.7.10
- OS: Ubuntu 21.10 (Linux-5.13.0-20-generic-x86_64-with-debian-11.0)
- (Optional) Other libraries and their versions:
```yaml
name: puma-lab
channels:
  - pyviz
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=1_gnu
  - abseil-cpp=20210324.2=h9c3ff4c_0
  - alembic=1.7.3=pyhd8ed1ab_0
  - alsa-lib=1.2.3=h516909a_0
  - anyio=3.3.0=py37h89c1867_0
  - argcomplete=1.12.3=pyhd8ed1ab_2
  - argon2-cffi=20.1.0=py37h5e8e339_2
  - arrow-cpp=5.0.0=py37hdf48254_5_cpu
  - async_generator=1.10=py_0
  - attrs=21.2.0=pyhd8ed1ab_0
  - autopage=0.4.0=pyhd8ed1ab_0
  - aws-c-cal=0.5.11=h95a6274_0
  - aws-c-common=0.6.2=h7f98852_0
  - aws-c-event-stream=0.2.7=h3541f99_13
  - aws-c-io=0.10.5=hfb6a706_0
  - aws-checksums=0.1.11=ha31a3da_7
  - aws-sdk-cpp=1.8.186=hb4091e7_3
  - babel=2.9.1=pyh44b312d_0
  - backcall=0.2.0=pyh9f0ad1d_0
  - backports=1.0=py_2
  - backports.functools_lru_cache=1.6.4=pyhd8ed1ab_0
  - backports.zoneinfo=0.2.1=py37h5e8e339_4
  - bleach=4.1.0=pyhd8ed1ab_0
  - bokeh=2.3.3=py37h89c1867_0
  - brotlipy=0.7.0=py37h5e8e339_1001
  - bzip2=1.0.8=h7f98852_4
  - c-ares=1.17.2=h7f98852_0
  - ca-certificates=2021.10.8=ha878542_0
  - certifi=2021.10.8=py37h89c1867_0
  - cffi=1.14.6=py37hc58025e_0
  - chardet=4.0.0=py37h89c1867_1
  - charset-normalizer=2.0.0=pyhd8ed1ab_0
  - click=8.0.1=py37h89c1867_0
  - clickhouse-cityhash=1.0.2.3=py37h3340039_2
  - clickhouse-driver=0.2.1=py37h5e8e339_0
  - cliff=3.9.0=pyhd8ed1ab_0
  - cloudpickle=2.0.0=pyhd8ed1ab_0
  - cmaes=0.8.2=pyh44b312d_0
  - cmd2=2.2.0=py37h89c1867_0
  - colorama=0.4.4=pyh9f0ad1d_0
  - colorcet=2.0.6=pyhd8ed1ab_0
  - colorlog=6.4.1=py37h89c1867_0
  - conda=4.10.3=py37h89c1867_1
  - conda-package-handling=1.7.3=py37h5e8e339_0
  - cramjam=2.3.1=py37h5e8e339_1
  - cryptography=3.4.7=py37h5d9358c_0
  - cycler=0.10.0=py_2
  - cytoolz=0.11.0=py37h5e8e339_3
  - dask=2021.9.0=pyhd8ed1ab_0
  - dask-core=2021.9.0=pyhd8ed1ab_0
  - datashader=0.13.0=pyh6c4a22f_0
  - datashape=0.5.4=py_1
  - dbus=1.13.6=h48d8840_2
  - debugpy=1.4.1=py37hcd2ae1e_0
  - decorator=5.1.0=pyhd8ed1ab_0
  - defusedxml=0.7.1=pyhd8ed1ab_0
  - distributed=2021.9.0=py37h89c1867_0
  - entrypoints=0.3=py37hc8dfbb8_1002
  - expat=2.4.1=h9c3ff4c_0
  - fastparquet=0.7.1=py37hb1e94ed_0
  - filelock=3.0.12=pyh9f0ad1d_0
  - fontconfig=2.13.1=hba837de_1005
  - freetype=2.10.4=h0708190_1
  - fsspec=2021.8.1=pyhd8ed1ab_0
  - gettext=0.19.8.1=h0b5b191_1005
  - gflags=2.2.2=he1b5a44_1004
  - gitdb=4.0.7=pyhd8ed1ab_0
  - gitpython=3.1.23=pyhd8ed1ab_1
  - glib=2.68.4=h9c3ff4c_0
  - glib-tools=2.68.4=h9c3ff4c_0
  - glog=0.5.0=h48cff8f_0
  - greenlet=1.1.1=py37hcd2ae1e_0
  - grpc-cpp=1.40.0=h850795e_0
  - gst-plugins-base=1.18.5=hf529b03_0
  - gstreamer=1.18.5=h76c114f_0
  - heapdict=1.0.1=py_0
  - holoviews=1.14.5=py_0
  - hvplot=0.7.3=py_0
  - icu=68.1=h58526e2_0
  - idna=3.1=pyhd3deb0d_0
  - importlib-metadata=4.8.1=py37h89c1867_0
  - importlib_metadata=4.8.1=hd8ed1ab_0
  - importlib_resources=5.2.2=pyhd8ed1ab_0
  - ipykernel=6.4.1=py37h6531663_0
  - ipympl=0.7.0=pyhd8ed1ab_0
  - ipython=7.27.0=py37h6531663_0
  - ipython_genutils=0.2.0=py_1
  - ipywidgets=7.6.5=pyhd8ed1ab_0
  - jbig=2.1=h7f98852_2003
  - jedi=0.18.0=py37h89c1867_2
  - jinja2=3.0.1=pyhd8ed1ab_0
  - joblib=1.0.1=pyhd8ed1ab_0
  - jpeg=9d=h36c2ea0_0
  - json5=0.9.5=pyh9f0ad1d_0
  - jsonschema=3.2.0=py37hc8dfbb8_1
  - jupyter-server-mathjax=0.2.3=pyhd8ed1ab_0
  - jupyter_client=7.0.2=pyhd8ed1ab_0
  - jupyter_contrib_core=0.3.3=py_2
  - jupyter_contrib_nbextensions=0.5.1=py37hc8dfbb8_1
  - jupyter_core=4.7.1=py37h89c1867_0
  - jupyter_highlight_selected_word=0.2.0=py37h89c1867_1002
  - jupyter_latex_envs=1.4.6=py37h89c1867_1001
  - jupyter_nbextensions_configurator=0.4.1=py37h89c1867_2
  - jupyter_server=1.11.0=pyhd8ed1ab_0
  - jupyterlab=3.1.11=pyhd8ed1ab_0
  - jupyterlab-git=0.32.2=pyhd8ed1ab_0
  - jupyterlab_pygments=0.1.2=pyh9f0ad1d_0
  - jupyterlab_server=2.8.1=pyhd8ed1ab_0
  - jupyterlab_widgets=1.0.2=pyhd8ed1ab_0
  - kiwisolver=1.3.2=py37h2527ec5_0
  - krb5=1.19.2=hcc1bbae_0
  - lcms2=2.12=hddcbb42_0
  - ld_impl_linux-64=2.36.1=hea4e1c9_2
  - lerc=2.2.1=h9c3ff4c_0
  - libarchive=3.5.2=hccf745f_0
  - libblas=3.9.0=11_linux64_openblas
  - libbrotlicommon=1.0.9=h7f98852_5
  - libbrotlidec=1.0.9=h7f98852_5
  - libbrotlienc=1.0.9=h7f98852_5
  - libcblas=3.9.0=11_linux64_openblas
  - libclang=11.1.0=default_ha53f305_1
  - libcurl=7.78.0=h2574ce0_0
  - libdeflate=1.7=h7f98852_5
  - libedit=3.1.20191231=he28a2e2_2
  - libev=4.33=h516909a_1
  - libevent=2.1.10=hcdb4288_3
  - libffi=3.3=h58526e2_2
  - libgcc-ng=11.1.0=hc902ee8_8
  - libgfortran-ng=11.1.0=h69a702a_8
  - libgfortran5=11.1.0=h6c583b3_8
  - libglib=2.68.4=h3e27bee_0
  - libgomp=11.1.0=hc902ee8_8
  - libiconv=1.16=h516909a_0
  - liblapack=3.9.0=11_linux64_openblas
  - libllvm11=11.1.0=hf817b99_2
  - libnghttp2=1.43.0=h812cca2_0
  - libogg=1.3.4=h7f98852_1
  - libopenblas=0.3.17=pthreads_h8fe5266_1
  - libopus=1.3.1=h7f98852_1
  - libpng=1.6.37=h21135ba_2
  - libpq=13.3=hd57d9b9_0
  - libprotobuf=3.16.0=h780b84a_0
  - libsodium=1.0.18=h36c2ea0_1
  - libsolv=0.7.19=h780b84a_5
  - libssh2=1.10.0=ha56f1ee_0
  - libstdcxx-ng=11.1.0=h56837e0_8
  - libta-lib=0.4.0=h516909a_0
  - libthrift=0.14.2=he6d91bd_1
  - libtiff=4.3.0=hf544144_1
  - libutf8proc=2.6.1=h7f98852_0
  - libuuid=2.32.1=h7f98852_1000
  - libuv=1.42.0=h7f98852_0
  - libvorbis=1.3.7=h9c3ff4c_0
  - libwebp-base=1.2.1=h7f98852_0
  - libxcb=1.13=h7f98852_1003
  - libxkbcommon=1.0.3=he3ba5ed_0
  - libxml2=2.9.12=h72842e0_0
  - libxslt=1.1.33=h15afd5d_2
  - llvmlite=0.37.0=py37h9d7f4d0_0
  - locket=0.2.0=py_2
  - lxml=4.6.3=py37h77fd288_0
  - lz4-c=1.9.3=h9c3ff4c_1
  - lzo=2.10=h516909a_1000
  - mako=1.1.5=pyhd8ed1ab_0
  - mamba=0.15.3=py37h7f483ca_0
  - markdown=3.3.4=pyhd8ed1ab_0
  - markupsafe=2.0.1=py37h5e8e339_0
  - matplotlib=3.4.3=py37h89c1867_0
  - matplotlib-base=3.4.3=py37h1058ff1_0
  - matplotlib-inline=0.1.3=pyhd8ed1ab_0
  - mistune=0.8.4=py37h5e8e339_1004
  - msgpack-python=1.0.2=py37h2527ec5_1
  - multipledispatch=0.6.0=py_0
  - mysql-common=8.0.25=ha770c72_2
  - mysql-libs=8.0.25=hfa10184_2
  - nb_conda_kernels=2.3.1=py37h89c1867_0
  - nbclassic=0.3.1=pyhd8ed1ab_1
  - nbclient=0.5.4=pyhd8ed1ab_0
  - nbconvert=6.1.0=py37h89c1867_0
  - nbdime=3.1.0=pyhd8ed1ab_0
  - nbformat=5.1.3=pyhd8ed1ab_0
  - ncurses=6.2=h58526e2_4
  - nest-asyncio=1.5.1=pyhd8ed1ab_0
  - notebook=6.4.3=pyha770c72_0
  - nspr=4.30=h9c3ff4c_0
  - nss=3.69=hb5efdd6_0
  - numba=0.54.0=py37h2d894fd_0
  - numpy=1.20.3=py37h038b26d_1
  - olefile=0.46=pyh9f0ad1d_1
  - openjpeg=2.4.0=hb52868f_1
  - openssl=1.1.1l=h7f98852_0
  - optuna=2.10.0=pyhd8ed1ab_0
  - orc=1.6.10=h58a87f1_0
  - packaging=21.0=pyhd8ed1ab_0
  - pandas=1.3.2=py37he8f5f7f_0
  - pandoc=2.14.2=h7f98852_0
  - pandocfilters=1.4.2=py_1
  - panel=0.12.1=py_0
  - param=1.11.1=pyh6c4a22f_0
  - parquet-cpp=1.5.1=1
  - parso=0.8.2=pyhd8ed1ab_0
  - partd=1.2.0=pyhd8ed1ab_0
  - patsy=0.5.2=pyhd8ed1ab_0
  - pbr=5.6.0=pyhd8ed1ab_0
  - pcre=8.45=h9c3ff4c_0
  - pexpect=4.8.0=py37hc8dfbb8_1
  - pickle5=0.0.11=py37h5e8e339_0
  - pickleshare=0.7.5=py37hc8dfbb8_1002
  - pillow=8.3.2=py37h0f21c89_0
  - pip=21.2.4=pyhd8ed1ab_0
  - prettytable=2.2.0=pyhd8ed1ab_0
  - prometheus_client=0.11.0=pyhd8ed1ab_0
  - prompt-toolkit=3.0.20=pyha770c72_0
  - psutil=5.8.0=py37h5e8e339_1
  - pthread-stubs=0.4=h36c2ea0_1001
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pyarrow=5.0.0=py37h58331f5_5_cpu
  - pycosat=0.6.3=py37h5e8e339_1006
  - pycparser=2.20=pyh9f0ad1d_2
  - pyct=0.4.6=py_0
  - pyct-core=0.4.6=py_0
  - pygments=2.10.0=pyhd8ed1ab_0
  - pykalman=0.9.5=py_1
  - pyopenssl=20.0.1=pyhd8ed1ab_0
  - pyparsing=2.4.7=pyh9f0ad1d_0
  - pyperclip=1.8.2=pyhd8ed1ab_2
  - pyqt=5.12.3=py37h89c1867_7
  - pyqt-impl=5.12.3=py37he336c9b_7
  - pyqt5-sip=4.19.18=py37hcd2ae1e_7
  - pyqtchart=5.12=py37he336c9b_7
  - pyqtwebengine=5.12.1=py37he336c9b_7
  - pyrsistent=0.17.3=py37h5e8e339_2
  - pysocks=1.7.1=py37h89c1867_3
  - python=3.7.10=hffdb5ce_100_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python_abi=3.7=2_cp37m
  - pytz=2021.1=pyhd8ed1ab_0
  - pyviz_comms=2.1.0=py_0
  - pyyaml=5.4.1=py37h5e8e339_1
  - pyzmq=22.2.1=py37h336d617_0
  - qt=5.12.9=hda022c4_4
  - re2=2021.09.01=h9c3ff4c_0
  - readline=8.1=h46c0cb4_0
  - redis-py=3.5.3=pyh9f0ad1d_0
  - reproc=14.2.3=h7f98852_0
  - reproc-cpp=14.2.3=h9c3ff4c_0
  - requests=2.26.0=pyhd8ed1ab_0
  - requests-unixsocket=0.2.0=py_0
  - ruamel_yaml=0.15.80=py37h5e8e339_1004
  - s2n=1.0.10=h9b69904_0
  - scikit-learn=0.24.2=py37hf0f1638_1
  - send2trash=1.8.0=pyhd8ed1ab_0
  - setproctitle=1.1.10=py37h5e8e339_1004
  - setuptools=58.0.4=py37h89c1867_0
  - six=1.16.0=pyh6c4a22f_0
  - smmap=3.0.5=pyh44b312d_0
  - snappy=1.1.8=he1b5a44_3
  - sniffio=1.2.0=py37h89c1867_1
  - sortedcontainers=2.4.0=pyhd8ed1ab_0
  - sqlalchemy=1.4.25=py37h5e8e339_0
  - sqlite=3.36.0=h9cd32fc_1
  - statsmodels=0.12.2=py37hb1e94ed_0
  - stevedore=3.4.0=py37h89c1867_0
  - ta-lib=0.4.19=py37ha21ca33_2
  - tabulate=0.8.9=pyhd8ed1ab_0
  - tblib=1.7.0=pyhd8ed1ab_0
  - tensorboardx=2.4=pyhd8ed1ab_0
  - terminado=0.12.1=py37h89c1867_0
  - testpath=0.5.0=pyhd8ed1ab_0
  - threadpoolctl=2.2.0=pyh8a188c0_0
  - thrift=0.13.0=py37hcd2ae1e_2
  - tk=8.6.11=h27826a3_1
  - toolz=0.11.1=py_0
  - tornado=6.1=py37h5e8e339_1
  - tqdm=4.62.2=pyhd8ed1ab_0
  - traitlets=5.1.0=pyhd8ed1ab_0
  - typing_extensions=3.10.0.0=pyha770c72_0
  - tzdata=2021a=he74cb21_1
  - tzlocal=3.0=py37h89c1867_2
  - urllib3=1.26.6=pyhd8ed1ab_0
  - wcwidth=0.2.5=pyh9f0ad1d_2
  - webencodings=0.5.1=py_1
  - websocket-client=0.57.0=py37h89c1867_4
  - wheel=0.37.0=pyhd8ed1ab_1
  - widgetsnbextension=3.5.1=py37h89c1867_4
  - xarray=0.19.0=pyhd8ed1ab_1
  - xeus=2.0.0=h7d0c39e_0
  - xeus-python=0.13.0=py37h4b46df4_1
  - xeus-python-shell=0.1.5=pyhd8ed1ab_0
  - xorg-libxau=1.0.9=h7f98852_0
  - xorg-libxdmcp=1.1.3=h7f98852_0
  - xz=5.2.5=h516909a_1
  - yaml=0.2.5=h516909a_0
  - zeromq=4.3.4=h9c3ff4c_1
  - zict=2.0.0=py_0
  - zipp=3.5.0=pyhd8ed1ab_0
  - zlib=1.2.11=h516909a_1010
  - zstandard=0.15.2=py37h5e8e339_0
  - zstd=1.5.0=ha95c52a_0
  - pip:
    - absl-py==0.13.0
    - aiohttp==3.7.4.post0
    - aiohttp-cors==0.7.0
    - aioredis==1.3.1
    - async-timeout==3.0.1
    - autograd==1.3
    - bayesian-optimization==1.2.0
    - blessed==1.19.0
    - blessings==1.7
    - cachetools==4.2.2
    - cma==2.7.0
    - colorful==0.5.4
    - cython==0.29.24
    - future==0.18.2
    - google-api-core==1.31.2
    - google-auth==1.35.0
    - google-auth-oauthlib==0.4.6
    - googleapis-common-protos==1.53.0
    - gpustat==0.6.0
    - gpy==1.10.0
    - gpytorch==1.5.1
    - grpcio==1.40.0
    - hebo==0.1.0
    - hiredis==2.0.0
    - multidict==5.1.0
    - nevergrad==0.4.3.post8
    - nvidia-ml-py3==7.352.0
    - oauthlib==3.1.1
    - opencensus==0.7.13
    - opencensus-context==0.1.2
    - paramz==0.9.5
    - protobuf==3.17.3
    - psycopg2==2.9.1
    - py-spy==0.3.9
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pydantic==1.8.2
    - pymoo==0.4.2.2
    - ray==1.7.1
    - requests-oauthlib==1.3.0
    - rsa==4.7.2
    - scipy==1.5.4
    - sklearn==0.0
    - tensorboard==2.6.0
    - tensorboard-data-server==0.6.1
    - tensorboard-plugin-wit==1.8.0
    - torch==1.9.1
    - werkzeug==2.0.1
    - yarl==1.6.3
```
### Steps to reproduce

1. run study
2. observe memory increasing on the node (the objective/tunable function consumes a constant amount)
3. OOM after a few hours

### Reproducible examples

```python
import optuna
import ray


@ray.remote
def optimize_optuna(df_data_in_object_store, labels_in_object_store, study_name, feature_name, pg_user, pg_pass,
                    pg_srv_ip):
    import optuna
    from feature import factory  # local module

    # disable all outputs
    optuna.logging.set_verbosity(optuna.logging.WARNING)
    optuna.logging.disable_default_handler()

    # this crashes the node with OOM when resuming from a "large" number of trials e.g. 100k (uses +32GB)
    # sampler = optuna.samplers.CmaEsSampler(use_separable_cma=True)

    study = optuna.create_study(
        study_name=study_name,
        direction="maximize",
        # sampler=sampler,
        load_if_exists=True,
        storage=f"postgresql://{pg_user}:{pg_pass}@{pg_srv_ip}/{study_name}")

    # get our trainable class from a local module
    trainable = factory.get_feature_cls(feature_name)(asks_in_object_store, labels_in_object_store)

    study.optimize(
        trainable, n_trials=None, timeout=None, gc_after_trial=False, show_progress_bar=False, 
        callbacks=None)

if __name__ == "__main__":
    # create the DB for a new study or do noting if already exist
    subprocess.run(
            ["psql", "-U", "myrole", "-d", "postgres", "-h", POSTGRESQL_SRV_IP, "-c", 
            f"CREATE DATABASE {study_name}"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL)
    
    # load data
    df_labeled = ...
    df_data = ...
    
    # connect to local ray cluster, propagate local env so we can call local 
    # modules from the remote function
    ray.init(
        runtime_env={"working_dir": "src", "excludes": ["*.txt", "bin", "doc", "examples", "notebooks"]}, 
        address="auto", 
        _redis_password="xxx")
    
    # put data in every node's shared memory so all processes running on that node can  
    # access a single object in memory (instead  of each process having it's own copy)
    ref_data = ray.put(df_data)
    ref_labeled = ray.put(df_labeled)
    
    resources_cluster = ray.cluster_resources()
    workers_available = int(resources_cluster["CPU"])

    # send tasks to the local cluster
    for _ in range(workers_available):
        optimize_optuna.remote(
            ref_data,
            ref_labeled,
            study_name,
            feature_name,
            constants.Env.POSTGRESQL_USER,
            db_pass,
            POSTGRESQL_SRV_IP
        )

    # block until the user interrupts
    while True:
        time.sleep(3600*24*31*356)
```

**Comment by himkt** [comment]
Hi @jmakov, sorry for the late response.

`CmaEsSampler` needs to fetch completed trials when sampling so I think it's not memory leak but data of tons of trials becomes too large (also, Optuna stores sampled parameters in JSON format and it also could be huge if you have many parameters to search).

If the problem also occurs with a sampler using the optimization history such as `TPESampler` and doesn't occur with a sampler that doesn't depend on the history such as `RandomSampler`, I think the behavior is not a bug.

**Comment by azagajewski** [comment]
Hi @himkt, it appears not to be a feature of the adaptive samplers - I just ran a hyperparameter optimization on a deep learning model using RandomSampler and MedianPruner, and the memory leak over trials is the same as TPESampler. It's a bit hard to reproduce since this ran for 3 hrs on a 4xGPU AWS node, but the rough code structure is as follows:


```
def objective(trial, strategy, **kwargs):
   
   keras.backend.clear_session()
   
   architecture = trial.suggest_categorical("arch", list(...)) 
   learning_rate = trial.suggest_float("learning_rate", low, high, log=True) 
   
   with strategy.scope():
      model = build_model(architecture, learning_rate)
      prune = [TFKerasPruningCallback(trial,'val_loss')]
      history = model.fit((callbacks=prune)
      return(history.history["val_loss"][-1]


study = optuna.create_study(direction="minimize",storage=db_URL, study_name=study_name, load_if_exists=True,sampler=optuna.samplers.RandomSampler(), pruner=optuna.pruners.MedianPruner(n_startup_trials=10,n_warmup_steps=3))

objective_wrapper = lambda trial: optuna_objective(trial, strategy=strategy)

study.optimize(objective_wrapper, n_trials=trials, gc_after_trial=True)
```

The `clear_session()` correctly flushes the GPU between trials and clears distribution settings, `with.strategy.scope()` recreates the distribution settings to use multiple GPUs, `gc_after_trial` reduces the memory leak but does not eliminate entirely. The RDB is a postgres on a remote server. 

Here is the diagnostic showing real and virtual memory increasing over the completion of several trials with the random sampler and median pruner - the process continues until the OS refuses to create new optuna threads due to the virtual memory being too high, leading to `LLVM ERROR: pthread_create failed: Resource temporarily unavailable`

![optuna_memory_leak](https://github.com/optuna/optuna/assets/44504386/6f1741ae-403e-42f8-93f7-687076a4d7c7)

**Comment by steveepreston** [comment]
I confirm this issue.
Memory in-use stacking trail by trail, until it raise out of memory error.

placing these inside objective() also not helped:
```python
K.clear_session()
gc.collect()
```

**Comment by himkt** [comment]
Hello @steveepreston. If `out of memory (OOM) error` means a CUDA error, `gc_after_trial` doesn't have any meaning since it enables a garbage collection on a CPU side. I think the problem is caused by not Optuna but something inside your objective function (I know, you did try `K.clear_session()` though). If you can share a simple script to replicate the problem (please make a script works without extra resource, since I need to run on my environment), I'll take a look.

side note; thank you for mentioning me in your comment. Currently I'm not an active developer and haven't noticed the discussion by @azagajewski, I'm sorry.

**Comment by steveepreston** [comment]
@himkt Thanks for attention!
no, i'm running on TPU VM.

the code:
```python
class RAMLogger(Callback):
    def on_epoch_begin(self, epoch, logs=None):
        print('RAM_used', psutil.virtual_memory().used / (1024 ** 3))


filters = 20
kernel_sizes = [3, 4, 5]
dropout_rate = 0.2
bert_model = "HooshvareLab/roberta-fa-zwnj-base"
batch_size = 256

tokenizer = AutoTokenizer.from_pretrained(bert_model, clean_up_tokenization_spaces=True)
train_dataset = tokenize_and_create_dataset(train_texts, train_labels, tokenizer, batch_size)
test_dataset = tokenize_and_create_dataset(test_texts, test_labels, tokenizer, batch_size)


def objective(trial):
    K.clear_session()
    gc.collect()

    with strategy.scope():
        print("Trial Start")

        model = BertCNN()
        model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
        model.fit(train_dataset, validation_data=test_dataset, epochs=5, verbose=model_verbose, callbacks=[RAMLogger()])

        predictions = model.predict(test_dataset, verbose=model_verbose)
        predicted_labels = np.argmax(predictions, axis=1)
        report = classification_report(test_labels, predicted_labels, output_dict=True)

        del model
        gc.collect()

        return report['accuracy']


study = optuna.create_study(direction='maximize', sampler=optuna.samplers.RandomSampler(seed=42))
study.optimize(objective, n_trials=30, gc_after_trial=True)
```

**Comment by steveepreston** [comment]
Thank you @himkt for attention.

code is here:
```python
bert_model = "bert-base-cased"
batch_size = 256

class RAMLogger(Callback):
    def on_epoch_begin(self, epoch, logs=None):
        print('RAM_used', psutil.virtual_memory().used / (1024 ** 3))


tokenizer = AutoTokenizer.from_pretrained(bert_model, clean_up_tokenization_spaces=True)
train_dataset = tokenize_and_create_dataset(train_texts, train_labels, tokenizer, batch_size)
test_dataset = tokenize_and_create_dataset(test_texts, test_labels, tokenizer, batch_size)


def objective(trial):
    K.clear_session()
    gc.collect()

    with strategy.scope():
        print("Trial Start")

        model = BertCNN()
        model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
        model.fit(train_dataset, validation_data=test_dataset, epochs=5, verbose=model_verbose, callbacks=[RAMLogger()])

        predictions = model.predict(test_dataset, verbose=model_verbose)
        predicted_labels = np.argmax(predictions, axis=1)
        report = classification_report(test_labels, predicted_labels, output_dict=True)

        del model
        gc.collect()

        return report['accuracy']


study = optuna.create_study(direction='maximize', sampler=optuna.samplers.RandomSampler(seed=42))
study.optimize(objective, n_trials=30, gc_after_trial=True)
```

**Comment by himkt** [comment]
Let me rephrase, `please make a script works without extra resource, since I need to run on my environment`. :pray:

And I think I saw `TPU` on your first comment. I just found https://stackoverflow.com/questions/74289376/keras-clear-session-not-working-in-google-colab, which said multiple `K.clear_session()` invocations didn't clear the memory. I want you to check if the problem is replicated without Optuna.

For example, I think we can check that by the following sample:

```python
def objective():
    K.clear_session()
    gc.collect()

    with strategy.scope():
        print("Trial Start")

        model = BertCNN()
        model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
        model.fit(train_dataset, validation_data=test_dataset, epochs=5, verbose=model_verbose, callbacks=[RAMLogger()])

        predictions = model.predict(test_dataset, verbose=model_verbose)
        predicted_labels = np.argmax(predictions, axis=1)
        report = classification_report(test_labels, predicted_labels, output_dict=True)

        del model
        gc.collect()

        return report['accuracy']


for _ in range(30):
    objective()
```

**Comment by akosmaroy** [comment]
I have the same issue, where I delete all objects created in the objective function, do a gc.collect(), but the memory footprint of the optuna script is still increasing with time.

Tried to measure the size of the trial object with pympler.asizeof() but this fails due to a weak reference error.

In this case, no TPU is involved, it's a CPU only optimization script.

**Comment by nzw0301** [comment]
@akosmaroy Could you share the minimal reproducible code? It sounds natural because `study`'s storage instance stores optimisation history: finished trials with default config.

**Comment by akosmaroy** [comment]
> @akosmaroy Could you share the minimal reproducible code? It sounds natural because `study`'s storage instance stores optimisation history: finished trials, with default config.

to grow to some extent, yes. but here the grows is in the 10s of Gigabytes on a couple of hours run.

didn't have the time to strip down the code to a small generic sample, when it comes to the optimize() function. this is how optuna is invoked:

```
    full_study_name = f"study-backtest"
    db_path = os.path.join(output_dir_path, f"{full_study_name}.db")

    directions = ['maximize', 'maximize']

    study = optuna.create_study(directions=directions,
                                storage=f"sqlite:///{db_path}",
                                study_name=full_study_name,
                                load_if_exists=True)
    study.optimize(objective_with_parameters, n_trials=trials, gc_after_trial=True, callbacks=[print_trial_callback_with_parameters])
```

in the objective function, I delete all locally created objects explicitly and call gc.collect(). I do the same in the callback.

**Comment by nzw0301** [comment]
Thanks, but I'm afraid that it's not possible to resolve your issue without the whole script.

**Comment by raphaelxjy** [comment]
> Hi @jmakov, sorry for the late response.
> 
> `CmaEsSampler` needs to fetch completed trials when sampling so I think it's not memory leak but data of tons of trials becomes too large (also, Optuna stores sampled parameters in JSON format and it also could be huge if you have many parameters to search).
> 
> If the problem also occurs with a sampler using the optimization history such as `TPESampler` and doesn't occur with a sampler that doesn't depend on the history such as `RandomSampler`, I think the behavior is not a bug.

Hi! Been experiencing a similar situation to the rest where I'm running 256 studies in parallel using Joblib on a 256 core system, and the objective function is pretty easy to compute, although the problem is complex (30+ dimension floats). I tried to run each study for 100000 trials but experienced out of memory errors (my system has about 250GB of RAM). Based on this response a few years back I understand that what I'm encountering seems to be natural behaviour of the system because there are simply too many trials, just wanted to reach out to get an updated response given that it's been a few years: Is this still the case? And is there anything I can do to work around what I'm experiencing, apart from running less processes in parallel to provide each one with more memory?

Thanks in advance!

**Comment by Furkan-rgb** [comment]
Came here, because I am running into the same issue. Turned on gc after trial: `study.optimize(objective, n_trials=max_trials, gc_after_trial=True, show_progress_bar=True)` and added K.clear_session() to no avail. Can't seem to figure out what is memory leaking.

**Comment by steveepreston** [comment]
I suggest that this issue re-opened, or because it's closed in 2021, Maybe a new issue is required.
Please link to here if you opened new issue.

**Comment by anskumar-akam** [comment]
I'm facing the same issue with tuning a pytorch model with only CPU.

**Comment by nzw0301** [comment]
In my experience, there are multiple reasons to cause OOM error and we need to [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to investigate it smoothly. Leaving the comment without it on this issue is less informative, so please open a new issue if you think the issue is a bug in the optuna's codebase  or ask a question on https://github.com/optuna/optuna/discussions. 

Best regards,
